compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -734,13 +733,36 @@ abstract class GenJSCode extends plugins.PluginComponent
         if (scalaPrimitives.isPrimitive(sym)) {
           None
         } else if (sym.isDeferred || sym.owner.isInterface) {
+          val body = if (sym.hasAnnotation(JavaDefaultMethodAnnotation)) {
+            /* For an interface method with @JavaDefaultMethod, make it a
+             * default method calling the impl class method.
+             */
+            val implClassSym = sym.owner.implClass
+            val implMethodSym = implClassSym.info.member(sym.name).suchThat { s =>
+              s.isMethod &&
+              s.tpe.params.size == sym.tpe.params.size + 1 &&
+              s.tpe.params.head.tpe =:= sym.owner.toTypeConstructor &&
+              s.tpe.params.tail.zip(sym.tpe.params).forall {
+                case (sParam, symParam) =>
+                  sParam.tpe =:= symParam.tpe
+              }
+            }
+            genTraitImplApply(implMethodSym,
+                js.This()(currentClassType) :: jsParams.map(_.ref))
+          } else {
+            js.EmptyTree
+          }
           Some(js.MethodDef(static = false, methodName,
-              jsParams, currentClassType, js.EmptyTree)(
+              jsParams, toIRType(sym.tpe.resultType), body)(
```

Comment:
I thought about that, and in fact it doesn't. But it's a limitation of our approach to IR checking. We only IR-check after linking (for good reasons, but still). But after linking, there are no abstract methods anymore, since abstract methods cannot be reached (by definition). Therefore, the IR checker never even sees these methods, and therefore does not error out on them.

Addressing this will require rethinking the IR checker strategy.